components/centraldashboard/public/components/utilities-mixin_test.js

```diff
@@ -1,37 +1,117 @@
 }
 
+const PASS_FUNCTION = (_) => _;
+
 describe('Utilities Mixin', () => {
     let testObject;
 
     beforeEach(() => {
         testObject = new TestClass();
+        // Since base class is not HTML
+        testObject.dispatchEvent = PASS_FUNCTION;
     });
 
     it('Correct evaluates a logical or expression', () => {
-        expect(testObject.or()).toBe(false);
-        expect(testObject.or(false, false, false, true)).toBe(true);
-        expect(testObject.or(false, false, false)).toBe(false);
-        expect(testObject.or(1, 0)).toBe(true);
-        expect(testObject.or('', 0)).toBe(false);
+        const {or} = testObject;
+        expect(or()).toBe(false);
+        expect(or(false, false, false, true)).toBe(true);
+        expect(or(false, false, false)).toBe(false);
+        expect(or(1, 0)).toBe(true);
+        expect(or('', 0)).toBe(false);
     });
 
     it('Correct evaluates a logical equals expression', () => {
-        expect(testObject.equals()).toBe(true);
-        expect(testObject.equals(1, '1')).toBe(false);
-        expect(testObject.equals(1, 1, 1)).toBe(true);
-        expect(testObject.equals('abc', 'abc', 'abc')).toBe(true);
-        expect(testObject.equals('abc', 'abc', 'ABC')).toBe(false);
+        const {equals} = testObject;
+        expect(equals()).toBe(true);
+        expect(equals(1, '1')).toBe(false);
+        expect(equals(1, 1, 1)).toBe(true);
+        expect(equals('abc', 'abc', 'abc')).toBe(true);
+        expect(equals('abc', 'abc', 'ABC')).toBe(false);
+    });
+
+    it('Checks empty-ness on objects correctly', () => {
+        const {empty} = testObject;
+        expect(empty()).toBe(true);
+        expect(empty(0)).toBe(true);
+        expect(empty({})).toBe(true);
+        expect(empty([])).toBe(true);
+        expect(empty(new Set([]))).toBe(true);
+
+        expect(empty(1)).toBe(false);
+        expect(empty({a: 1})).toBe(false);
+        expect(empty([1])).toBe(false);
+        expect(empty(new Set([1]))).toBe(false);
+        expect(empty(new Error())).toBe(false);
+        expect(empty(new CustomEvent('test'))).toBe(false);
     });
 
     it('Builds links with valid query parameters', () => {
-        expect(testObject.buildHref('/foo', null)).toBe('/foo');
-        expect(testObject.buildHref('/foo?other_param=value', {}))
+        const {buildHref} = testObject;
+        expect(buildHref('/foo', null)).toBe('/foo');
+        expect(buildHref('/foo?other_param=value', {}))
             .toBe('/foo?other_param=value');
-        expect(testObject.buildHref('/foo?other_param=value', {ns: 'bar'}))
+        expect(buildHref('/foo?other_param=value', {ns: 'bar'}))
             .toBe('/foo?other_param=value&ns=bar');
     });
+
+    it('FireEvent correctly fires Event-Obj or custom events', () => {
+        const fireEvent = testObject.fireEvent.bind(testObject);
+        let lastEvent;
+        spyOn(testObject, 'dispatchEvent').and.callFake((ev) => {
```

Comment:
There's no need to provide the fake implementation of `dispatchEvent`. The `fireEvent` method's logic builds the argument for `dispatchEvent` based on its parameters so all you really need to do is call `fireEvent` and then validate that the mock was called with the arguments you expect.

Reply:
Dispatch event gets called with an event object, how would it compare to event objects?
- The event I construct to compare with
- The event generated by `fireEvent` that is fired?

Reply:
Here's roughly the implementation I would use. The purpose of this test is to verify the interaction with `dispatchEvent`, which is not part of your implementation but is provided by the CustomElement class utilizing the mixin. So you create the spy function for `dispatchEvent` and verify that it gets called with the type of object you expect. The end result is similar to what you started with, but you don't need to have any kind of fake implementation since you're only verifying how dispatchEvent was called. Normally, you could just use the `toHaveBeenCalledWith` matcher, but it doesn't work well when comparing Event objects.

```
let lastEvent;
testObject.dispatchEvent = jasmine.createSpy();
testObject.fireEvent('foo');

lastEvent = testObject.dispatchEvent.calls.mostRecent().args[0];
expect(lastEvent.type).toBe('foo');
expect(lastEvent.detail).toEqual(null);

testObject.fireEvent('bar', {ap: 1});
lastEvent = testObject.dispatchEvent.calls.mostRecent().args[0];
expect(lastEvent.type).toBe('bar');
expect(lastEvent.detail).toEqual({ap: 1});

const ev = new Event('click');
testObject.fireEvent(ev);
lastEvent = testObject.dispatchEvent.calls.mostRecent().args[0];
expect(lastEvent).toBe(ev);
```

Reply:
Yup done!